### PR TITLE
Issue842 multiyear

### LINKED ISSUE #842: ReaderTMY3 does not work for data going over into the next year
Hello everyone,

I have my own weather data, which I have adapted to a TMY3 format, in order to use the ReaderTMY3. The data covers the period of June 2016 to October 2017. The time stamp I used in the TMY3 format sets 1.01.2016 00:00 to 0s. However I get an error at 01.01.2017 00:00 and the simulation stops. Is there a way around this, as depending on my application, I might have data stretching over more than a year anyway? Or am not setting some parameter correctly?
Otherwise I'm happy to assist in extending the ReaderTMY3 to deal with such a situation.

### PR DESCRIPTION
**Pull request by mwetter**
This closes #842 

Todo:
- [x] The test `BoundaryConditions/WeatherData/BaseClasses/Examples/ConvertTime.mo` fails as the model triggers an assertion when advancing to the next year.

@AnaConstantin: Can you please check the logic in `BoundaryConditions/WeatherData/BaseClasses/ConvertTime.mo`. I don't think the above test should fail.

### CODE REVIEW

**Comment by mwetter** [comment]
- [ ] The `getTimeSpanTMY3` function should be changed so that it returns for end time the last time stamp of the weather data plus the time increment, because this is the time span for a weather data.
Then, `ConvertTime` can first check if end time minus start time is a whole year (31536000 seconds) and if so, use the `mod` function. If it is not a whole year, then it should use an assertion if `time > weaDatEndTime`.
This way, we can localize the decision of how long the weather data are, and the `ConvertTime` block simply needs to convert the time and not estimate (with `weaDatAveInc`) how the weather data may be. The parameter `weaDatAveInc` can then be removed as it will only be used inside `getTimeSpanTMY3`.

**Comment by AnaConstantin** [comment]
@mwetter I think I get it. I'll go over the documentation and code comments to make sure endTime = lastTimeStamp + AvgIncrement is highlighted enough. I do not think this is the most intuitive way to do express endTime, but it's consistent with the rest of the code. Just one question: do you want to check only for a year, or also for multiples of a year, e.g. weather data stretching over two years?

**Comment by mwetter** [comment]
@AnaConstantin : We should allow for multiple years of weather data. I believe all the information is available to do the check in the `initial equation` section.

**Comment by mwetter** [comment]
@AnaConstantin @mlauster @Mathadon 
This is ready for review. I added more validation cases and updated the documentation.
Note that models that use the weather data reader such as the `ThermalZones.ReducedOrder` have a small difference at some time steps. I believe this is due to a change in numerics. I see similar changes in our larger models in the `Buildings` library. 
Please review and approve/require changes for this pull request so that we can close it.

**Comment by Mathadon** [comment]
[image1.jpeg]
Working on it, but first things first! Greetings from the modelica conference!

**Comment by AnaConstantin** [comment]
@mwetter thank you for the changes. To me it looks very good. My only question is if the assert has to be outside of the when loop.

**Comment by mwetter** [comment]
@AnaConstantin : Why do you think it should be inside the `when` clause? The conditions `(time - weaDatEndTim) < shiftSolarRad` and `when modTim > pre(tNext) then` are different. If you propose to move the assertion inside the `when` clause, then we should probably use the same conditions and test to make sure it is correct. Did you look into it?

**Comment by AnaConstantin** [comment]
@mwetter you are right. I didn‘t think it through. This when clause is better than the older versions, since it gets activated only when tNext has to be changed. So the condition in the assert is totally separate. Next time I‘ll take your advice and verify my assumptions before stating them. Sorry about that.

File: IBPSA/BoundaryConditions/WeatherData/BaseClasses/ConvertTime.mo [hunk 1]
```diff
@@ -12,28 +16,57 @@ block ConvertTime
...
-  constant Modelica.SIunits.Time year=31536000 "Number of seconds in a year";
-  discrete Modelica.SIunits.Time tStart "Start time of period";
+  constant Modelica.SIunits.Time shiftSolarRad=1800 "Number of seconds for the shift for solar radiation calculation";
+  parameter Modelica.SIunits.Time lenWea = weaDatEndTim-weaDatStaTim "Length of weather data";
+
+  parameter Boolean canRepeatWeatherFile = abs(mod(lenWea, 365*24*3600)) < 1E-2
+    "true if the weather file can be repeated, since it has the lenth of a year or a multiple of it";
 
+  Modelica.SIunits.Time tNext "Start time of next period";
+  function getNextTime
```
Comment (Mathadon) [inline IBPSA/BoundaryConditions/WeatherData/BaseClasses/ConvertTime.mo]:
```suggestion
  function getNextTime "Function that computes the next time when a switch needs to happen"
```

File: IBPSA/BoundaryConditions/WeatherData/BaseClasses/ConvertTime.mo [hunk 2]
```diff
@@ -12,28 +16,57 @@ block ConvertTime
...
-    tStart = integer(modTim/year)*year;
+  when modTim > pre(tNext) then
+    // simulation time stamp went over the end time of the weather file
+    //(last time stamp of the weather file + average increment)
+    tNext = if canRepeatWeatherFile then getNextTime(modTim, lenWea) else integer(Modelica.Constants.inf);
   end when;
-  calTim = modTim - tStart;
+  calTim = if canRepeatWeatherFile then modTim - (tNext - lenWea) else modTim;
+  assert(canRepeatWeatherFile or (time - weaDatEndTim) < shiftSolarRad,
+    "Insufficient weather data provided for the desired simulation period.",
```
Comment (Mathadon) [inline IBPSA/BoundaryConditions/WeatherData/BaseClasses/ConvertTime.mo]:
```suggestion
    "In " + getInstanceName() + ": Insufficient weather data provided for the desired simulation period.",
```

File: IBPSA/BoundaryConditions/WeatherData/BaseClasses/getLatitudeTMY3.mo [hunk 3]
```diff
@@ -14,7 +14,7 @@ algorithm
     IBPSA.BoundaryConditions.WeatherData.BaseClasses.getHeaderElementTMY3(
       filNam=filNam,
       start="#LOCATION",
-      name=  "longitude",
+      name = "latitude",
```
Comment (Mathadon) [inline IBPSA/BoundaryConditions/WeatherData/BaseClasses/getLatitudeTMY3.mo]: looks scary, but apparently this is just for error reporting!

File: IBPSA/BoundaryConditions/WeatherData/ReaderTMY3.mo [hunk 4]
```diff
@@ -1452,9 +1460,28 @@ and allows the following configurations:
   </td>
 </tr>
 </table>
+<!-- ============================================== -->
+<h4>Length of weather data and simulation period</h4>
+<p>
+If weather data span a year, which is the default for TMY3 data, or multiple years,
+then this model can be used for simulations that span multiple years. The simulation
+start time need to be set to the clock time of the respective start time. For example,
```
Comment (Mathadon) [inline IBPSA/BoundaryConditions/WeatherData/ReaderTMY3.mo]:
```suggestion
start time needs to be set to the clock time of the respective start time. For example,
```

File: IBPSA/BoundaryConditions/WeatherData/ReaderTMY3.mo [hunk 5]
```diff
@@ -1452,9 +1460,28 @@ and allows the following configurations:
...
+to start at January 2 at 10am, set start time to <code>t=(24+10)*3600</code> seconds.
```
Comment (Mathadon) [inline IBPSA/BoundaryConditions/WeatherData/ReaderTMY3.mo]: I don't think this is sufficiently specific/clear. This assumes that the timezone that you define in the TMY3 file corresponds to the time zone of the date that you mention in the documentation?

Comment (mwetter) [reply IBPSA/BoundaryConditions/WeatherData/ReaderTMY3.mo]: @Mathadon : Not sure what you mean by not being clear. Can you add/propose better wording? Note that the weather files for Chicago and San Francisco both start at `0`. Hence I am not sure what you mean.

Comment (Mathadon) [inline IBPSA/BoundaryConditions/WeatherData/ReaderTMY3.mo]:
```suggestion
to start at January 2 at 10am, set start time to <code>t=(24+10)*3600</code> seconds.
For this computation, the used date and time (here January 2, 10 am) must be expressed in the same time zone
as the one that is used to define the TMY3 file. This is usually the local (winter) time zone.
The parameter `timZon` represents the TMY3 file time zone, expressed in seconds compared to UTC.
```
@mwetter does this formulation clarify what I mean?

**Comment by mwetter** [comment]
@Mathadon @AnaConstantin 
Thanks for the feedback, and the clearer wording from Filip. I will merge it after the tests pass.

**Pull request merged**